plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeTransactionLogEntryIterator.java

```diff
@@ -0,0 +1,121 @@
+
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.Optional;
+
+import static com.google.common.base.Verify.verify;
+import static io.trino.plugin.deltalake.DeltaLakeErrorCode.DELTA_LAKE_INVALID_SCHEMA;
+import static io.trino.plugin.deltalake.transactionlog.TransactionLogParser.readLastCheckpoint;
+import static io.trino.plugin.deltalake.transactionlog.TransactionLogUtil.getTransactionLogDir;
+import static io.trino.plugin.deltalake.transactionlog.TransactionLogUtil.getTransactionLogJsonEntryPath;
+import static io.trino.plugin.deltalake.transactionlog.checkpoint.TransactionLogTail.getEntriesFromJson;
+import static java.util.Objects.requireNonNull;
+
+public class DeltaLakeTransactionLogEntryIterator
+        implements Iterator<DeltaLakeTransactionLogEntry>
+{
+    private final TrinoFileSystem fileSystem;
+    private final Path transactionLogDir;
+    private long currentVersion;
+    private Optional<Iterator<DeltaLakeTransactionLogEntry>> currentEntries = Optional.empty();
+    private final long startVersion;
+    private final Optional<Long> endVersion;
+
+    private final Optional<Long> lastCheckpointVersion;
+
+    public DeltaLakeTransactionLogEntryIterator(
+            TrinoFileSystem fileSystem,
+            Path tableLocation,
+            Optional<Long> startVersion,
+            Optional<Long> endVersion)
+    {
+        verify(startVersion.orElse(0L) <= endVersion.orElse(startVersion.orElse(0L)), "startVersion must be less or equal than endVersion");
+        currentVersion = startVersion.orElse(0L) - 1;
+        this.startVersion = startVersion.orElse(0L);
+        this.endVersion = endVersion;
+        this.fileSystem = requireNonNull(fileSystem, "fileSystem is null");
+        transactionLogDir = getTransactionLogDir(requireNonNull(tableLocation, "tableLocation is null"));
+        lastCheckpointVersion = readLastCheckpoint(fileSystem, tableLocation).map(LastCheckpoint::getVersion);
+    }
+
+    @Override
+    public boolean hasNext()
+    {
+        if (currentEntries.isPresent()) {
+            if (currentEntries.get().hasNext()) {
+                return true;
+            }
+            else {
+                if (endVersion.isPresent() && (currentVersion == endVersion.get())) {
+                    return false;
+                }
+            }
+        }
+        if (currentVersion >= startVersion &&
+                currentEntries.isEmpty() &&
+                (lastCheckpointVersion.isEmpty() || currentVersion >= lastCheckpointVersion.get())) {
+            // Do not retrieve the next transaction log entry when the current one is missing
+            return false;
+        }
+
+        while (true) {
+            // Retrieve next transaction log entry
+            currentVersion++;
+            try {
+                currentEntries = getEntriesFromJson(currentVersion, transactionLogDir, fileSystem)
```

Comment:
Aside from `commitInfo`, `getEntriesFromJson` also loads other entries. Can we load only `commitInfo`?

Reply:
https://github.com/trinodb/trino/pull/16192#discussion_r1118333876